lib/SILOptimizer/Utils/CanonicalizeInstruction.cpp

```diff
@@ -64,16 +95,121 @@ simplifyAndReplace(SILInstruction *inst, CanonicalizeInstruction &pass) {
       [&pass](SILInstruction *deleted) { pass.killInstruction(deleted); });
 
   // Push the new instruction and any users onto the worklist.
-  pass.newUsers(result);
+  pass.notifyHasNewUsers(result);
   return nextII;
 }
+
+//===----------------------------------------------------------------------===//
+//                        Canonicalize Memory Operations
+//===----------------------------------------------------------------------===//
+
+// Given a load with multiple struct_extracts/tuple_extracts and no other uses,
+// canonicalize the load into several (struct_element_addr (load)) pairs.
+//
+// (struct_extract (load %base))
+//   ->
+// (load (struct_element_addr %base, #field)
+static SILBasicBlock::iterator
+splitAggregateLoad(LoadInst *loadInst, CanonicalizeInstruction &pass) {
```

Comment:
Should this also handle load_borrow? I am not sure what that would mean in terms of the structuring. Just some food for thought.

Reply:
Possibly, depending on where SILGen emits LoadBorrow. But not in this 5.1 PR. I added a TODO here.